hack/run-hive-locally.sh

```diff
@@ -0,0 +1,363 @@
+#!/usr/bin/env bash
+
+# TODO: hiveadmission support
+# TODO: run hive-controllers/hive-clustersync separately with multiple replicas each (need docker to avoid port collisions)
+
+set -e
+set -u
+set -o pipefail
+
+# Highlighted echo
+function hl_echo {
+    tput setaf 3
+    echo "$@"
+    tput sgr0
+}
+
+function usage {
+    echo "The script runs a Hive component ($hive_component_operator or $hive_component_controllers) locally
+against the active cluster, simulating in-cluster environment. It captures the in-cluster Hive
+configurations, scales down the selected component, then runs it locally with the same environment.
+On pressing ctrl + c, the script is terminated and relevant components are scaled back up.
+
+Warnings:
+- The active cluster might be in use, in which case the script could interfere with its functioning.
+
+Notes:
+- Artifacts produced are stored in the --artifacts_folder which is removed once the script ends unless
+--preserve-artifacts is specified.
+- Selecting $hive_component_controllers runs all Hive controllers (including $hive_component_clustersync) together
+as a single replica.
+- Endpoints (e.g. metrics, pprof) of the Hive component running locally are directly exposed on localhost.
+
+Usage:
+- ./hack/run-hive-locally.sh COMPONENT [--flag=value ...]
+
+Parameters:
+- COMPONENT: must be either $hive_component_operator or $hive_component_controllers
+
+Optional flags:
+- --hive-operator-ns=hive: namespace where hive-operator is installed.
+- --log-level=debug: log level of the component to run.
+- --artifacts-folder=artifacts: name of the folder that will be created for artifacts.
+- --preserve-artifacts: if specified, preserve the artifacts folder after the script ends.
+- --help"
+}
+
+function kill_background_jobs {
+    local pid
+    while read -r pid; do
+        hl_echo "Killing process $pid"
+        kill "$pid" || true
+    done < <(jobs -p)
+    wait
+}
+
+function get_envs {
+    local type_name="$1"
+    local ns="$2"
+    # Arrange output into the following format:
+    # name1 value1
+    # name2 value2
+    # ...
+    local jsonpath_query='{range .spec.template.spec.containers[0].env[*]}{@.name} {@.value}{"\n"}{end}'
+    local env
+    local value
+    local resolved_value
+
+    hl_echo "Getting environment variables from $type_name in ns/$ns"
+    while read -r env value; do
+        # Skip environment variables that are already resolved
+        if [ ${resolved_envs["$env"]+_} ]; then
+            hl_echo "Environment variable $env is already resolved, skipping"
+            continue
+        fi
```

Comment:
You are right. I also hesitated between running them separately (in Docker or so to avoid port collision) and together. Later I started to think that the script is meant to be a simple, quick way of debugging etc. with few dependencies, so in the end I stuck with the current implementation and added a TODO at the top of the script.

Reply:
> running them separately (in Docker or so to avoid port collision)

It's a pretty thin line between running in Docker and just running in KIND, which is already possible (or so they tell me -- I've never done it).

But yeah, I understand the issue. Hopefully it won't be a problem.

Reply:
This reminds me about a comparison I made between approaches for running/developing/testing Hive:
1) Openshift cluster + remote image registry (quay.io for ex.): 
Closest to production environment; 
Cluster creation is slow;
Requires an image build (so implicit requirements on platform architecture) each time the source code is updated; 
Image pushing/pulling is slow; 
This is QEs' go-to approach. 
2) KIND cluster + local image build/push/pull using `./hack/create-kind-cluster.sh`: 
Kubernetes/containerized environment; 
cluster creation is fast;
Requires an image build each time the source code is updated;  
Image pushing/pulling is fast; 
Has some requirements on memory, dependencies, setups (`./hack/hiveadmission-dev-cert.sh`) etc. 
3) Openshift + script merged by this PR: 
Runs bare components locally;
Cluster creation is slow;
Does not require image building/pushing/pulling at all;
Has minimal requirements.

Reply:
> 1. Requires an image build

Note that there's a little shortcut the dev team uses. It does depend on your host OS being fedora/RHEL:
`make image-hive-fedora-dev-base` creates the slow and seldom-changing parts (including `dnf update`) of the docker image.
Then `make docker-dev-push` knows to look for the image generated by the above and add just the hive bits on top.
(Pushing is still slow, yeah.)

Reply:
Would be great to save some time with this. 

-- Me on my Mac with an ARM chip :sob: